File: battleship_refree.py
import parkutil
import battleship

class battleship_refree:
    def __init__(self, w, h):
        parkutil.register_function('deploy', 2)
        parkutil.register_function('play', 4)
        self.players = parkutil.get_players(2)

        self.skipped = 0
        self.turn = True # player 0 goes first
        self.board_dim = (w, h)
        self.boards = [battleship.board(*self.board_dim) for i in range(len(self.players))]

    def start(self):
        deployed = [self.deploy(i) for i in range(len(self.players))]
        while True:
            result = self.over()
            if result == 2:
                print ('Game ended with a draw')
                break
            elif result != -1:
                print ('%s won!' % parkutil.player_uid(result))
                break

            self.turn = not self.turn
            self.playturn()

    def over(self):
        # draw cases
        if self.boards[0].lost and self.boards[1].lost:
            return 2
        if self.boards[0].skipped() and self.boards[1].skipped():
            return 2

        # only need to check the opponent
        if self.boards[not self.turn].lost:
            return int(self.turn)
        return -1 # nobody won yet

    def playturn(self):
        m = self.boards[self.turn]
        o = self.boards[not self.turn]
        h = None
        try:
            h = self.players[self.turn].play(m.get(), o.hidden(), m.last_attacked)
        except ClockTimeout:
            pass
        finally:
            o.attack(h)

    def deploy(self, i):
        try:
            ships = self.players[i].deploy(*self.board_dim)
            self.boards[i].deploy(ships)
        except ClockTimeout:
            self.boards[i].lost = True

if __name__ == '__main__':
    game = battleship_refree(10, 10)
    game.start()

File: battleship_refree.rb
require_relative 'parkutil'
require_relative 'battleship'

class BattleshipRefree

  def initialize(w, h)
    Parkutil.register_class(Battleship::Block)
    Parkutil.register_function('deploy', 2)
    Parkutil.register_function('play', 4)
    Parkutil.load_players(2)

    @board_dim = [w, h]
    @boards = Parkutil.count_players.times.map do |i|
      Battleship::Board.new(*@board_dim)
    end
  end

  def start
    Parkutil.count_players.times { |i| self.deploy(i) }

    while true do
      self.over!
      self.playturn
    end
  end

  def over!
    # draw
    exit 1 if @boards.all? { |b| b.lost } or @boards.all? { |b| b.opp_skipped }

    # check myself if lost
    if @boards[Parkutil.turn].lost then
      puts Parkutil.player(Parkutil.turn(-1)).uid
      exit 0
    end
  end

  def playturn
    m = @boards[Parkutil.turn]
    o = @boards[Parkutil.turn 1]
    h = nil

    begin
      h = Parkutil.current_player.play(m.get, o.hidden, m.last_attacked, o.last_sunk)
    rescue Parkutil::ClockTimeout
    rescue => e
      puts Parkutil.current_player.uid
      Parkutil.print_exception(e)
      exit 2
    end

    o.attack h
    Parkutil.advance_turn
  end

  def deploy(i)
    begin
      ships = Parkutil.player(i).deploy(*@board_dim)
      @boards[i].deploy(ships)
    rescue Parkutil::ClockTimeout
      @boards[i].lost = true
    rescue => e
      puts Parkutil.player(i).uid
      Parkutil.print_exception(e)
      exit 2
    end
  end

end

BattleshipRefree.new(10, 10).start()
